### PR TITLE
Implement deployment and verification steps as state machine operations.

### PR DESCRIPTION
**Pull request by simoncochrane**

### CODE REVIEW

File: gok8s/workload/provider.go [hunk 1]
```diff
@@ -41,93 +37,94 @@ type Resource struct {
...
+	return k.cs
 }
 
-// SyncWorkload checks if container version of all workload that matches CV selector
-// is up to date with the version that is requested, if not it
-// performs a rollout with specified roll-out strategy on deployment.
-func (k *K8sProvider) SyncWorkload(cv *cv1.ContainerVersion, version string) error {
-	specs, err := k.getMatchingWorkloadSpecs(cv)
+// CanRollout returns true if we can attempt a rollout for the cv resource at the given version.
+func (k *Provider) CanRollout(cv *cv1.ContainerVersion, version string) bool {
```
Comment (suneeta-mall) [inline gok8s/workload/provider.go]: I am not sure if the CanRollout interface should be as simple as that .. I think Target is a big part of whether rollout was successful or not ... for ex. may be container is perfectly valid but the config on target container is causing one target to fail but other targets are fine.. do we treat that as failed rollout version?

Comment (suneeta-mall) [inline gok8s/workload/provider.go]: I am not sure if CanRollout check should be as simple as that .. target can play a big role in it . ex bad config on one of target thats causing failed on that specific one but other targets matching cv specs are okay

File: sync/sync.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		version, err := s.registry.Version(s.cv.Spec.Tag)
+		if err != nil {
+			s.options.Recorder.Event(events.Warning, "CRSyncFailed", "Failed to get version from registry")
+			return state.Error(errors.Wrap(err, "failed to get version from registry"))
+		}
+
+		log.Printf("Current version: %v", version)
+
+		if !s.k8sProvider.CanRollout(s.cv, version) {
+			log.Printf("Not attempting %s rollout of version %s: already failed.", s.cv.Name, version)
```
Comment (suneeta-mall) [inline sync/sync.go]: Not sure about this check .. this should have more to do container verification I would think

File: sync/sync.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+			return state.None()
+		}
+
```
Comment (suneeta-mall) [inline sync/sync.go]: Talking about verification .. sounds like the container verification step is lost in refactor ? It should at this level I would think now .. before we start looking at matching workload

File: state/machine.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	}
+
+	if failureFunc != nil {
+		newOp.failureFuncs = append(newOp.failureFuncs, failureFunc)
+	}
+	return newOp
+}
+
+// Machine implements the main state machine loop.
+type Machine struct {
```
Comment (suneeta-mall) [inline state/machine.go]: I am not fan of calling it state machine .. its more pipeline .. we have parallel or single pipeline. State machine is has various state but at one point it has only one valid state ... 

I would prefer if state is more represented as step/stage or someone more alike and machine as pipeline (a better word may be).

File: sync/sync.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+				toUpdate = append(toUpdate, wl)
+			}
+		}
+
+		log.Printf("Found %d workloads to update", len(toUpdate))
+
+		var states []state.State
+		for _, wl := range toUpdate {
+			st := verify.NewVerifiers(s.k8sProvider.Client(), s.k8sProvider.Namespace(), version, s.cv.Spec.Container.Verify,
+				deploy.NewDeployState(s.k8sProvider.Client(), s.k8sProvider.Namespace(), s.cv, version, wl,
```
Comment (suneeta-mall) [inline sync/sync.go]: I dont see the config options passed in at required places .. did I miss them?

File: deploy/bluegreen.go [hunk 6]
```diff
@@ -1,15 +1,14 @@
...
+	"context"
 	"log"
 	"strings"
 	"time"
 
-	"github.com/nearmap/cvmanager/events"
 	cv1 "github.com/nearmap/cvmanager/gok8s/apis/custom/v1"
 	"github.com/nearmap/cvmanager/history"
-	"github.com/nearmap/cvmanager/stats"
+	"github.com/nearmap/cvmanager/state"
```
Comment (suneeta-mall) [inline deploy/bluegreen.go]: Blue green does not handle rollback still .. I would think its easier to do that now .. ?

File: deploy/simple.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+func (sd *SimpleDeployer) checkRollbackState(container *v1.Container, next state.State) state.StateFunc {
+	return func(ctx context.Context) (state.States, error) {
+		if sd.target.RollbackAfter() == nil {
+			log.Printf("Target %s does not define a progress deadline.", sd.target.Name())
+			return state.Single(next)
+		}
+
+		log.Printf("Checking progress health of target %s with deadline %v", sd.target.Name(), sd.target.RollbackAfter())
+
+		prevVersion := strings.SplitAfterN(container.Image, ":", 2)[1]
```
Comment (suneeta-mall) [inline deploy/simple.go]: I am inclined to put this on target/pod interface and give container image and version based on name of container provided. Would save from from this ugly split all over the places that we are doing (not related to your PR but just a thought)

**Comment by simoncochrane** [comment]
Superceded by https://github.com/nearmap/cvmanager/pull/34.

**Pull request closed without merging**